[HttpKernel] Caching HttpKernel instance in Kernel

gnugat opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | N/A |
| License | MIT |
| Doc PR | N/A |

After some profiling I've spotted that `Kernel` was calling `$container->get('http_kernel')` 3 times when using:

``` php
<?php

use Symfony\Component\HttpFoundation\Request;

require_once __DIR__.'/vendor/autoload.php';

$kernel = new AppKernel('prod', true);
$kernel->loadClassCache();
Request::enableHttpMethodParameterOverride();
$request = Request::createFromGlobals();
$response = $kernel->handle($request); // calling `$container->get('http_kernel')` 1 time 
$response->send();
$kernel->terminate($request, $response); // calling `$container->get('http_kernel')` 2 times 
```

It's mostly fine because `http_kernel` is a shared service (so `HttpKernel` is instantiated only once), but maybe we could still reduce those calls to 1 by saving the instance in a class property?
There's actually many ways to do this (remove `getHttpKernel()`, or add the property initialization in `getHttpKernel()`, etc), so here's a first patch.

[See profiling comparison](https://blackfire.io/profiles/compare/7f6b2e9f-d0ac-48bd-ae61-9270e1e18773/graph)

gnugat committed:
[HttpKernel] Caching HttpKernel instance in Kernel

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | N/A

After some profiling I've spotted that `Kernel` was calling `$container->get('http_kernel')` 3 times when using:

```php
<?php

use Symfony\Component\HttpFoundation\Request;

require_once __DIR__.'/vendor/autoload.php';

$kernel = new AppKernel('prod', true);
$kernel->loadClassCache();
Request::enableHttpMethodParameterOverride();
$request = Request::createFromGlobals();
$response = $kernel->handle($request); // calling `$container->get('http_kernel')` 1 time 
$response->send();
$kernel->terminate($request, $response); // calling `$container->get('http_kernel')` 2 times 
```

It's mostly fine because `http_kernel` is a shared service (so `HttpKernel` is instantiated only once), but maybe we could still reduce those calls to 1 by saving the instance in a class property?
There's actually many ways to do this (remove `getHttpKernel()`, or add the property initialization in `getHttpKernel()`, etc), so here's a first patch.

[See profiling comparison](https://blackfire.io/profiles/compare/7f6b2e9f-d0ac-48bd-ae61-9270e1e18773/graph)

xabbuh commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -166,7 +169,7 @@ public function handle(Request $request, $type = HttpKernelInterface::MASTER_REQ
         }
 
-        return $this->getHttpKernel()->handle($request, $type, $catch);
+        return $this->httpKernel>handle($request, $type, $catch);
```
This now breaks when the kernel was not booted before (see the failing test).

gnugat replied:
Oh, I would assume that the `if` statement just above would prevent this case

gnugat replied:
Actually it was just a typo (`>` instead of `->`)

xabbuh commented:
Status: Needs work

gnugat committed:
Fixed typos

sstok commented:
The property must be set to null when shutting down and when cloning.

Looking at the profiler comparison the gain seems to be very minimum.

stof commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -58,6 +58,7 @@
     protected $name;
     protected $startTime;
     protected $loadClassCache;
+    protected $httpKernel;
```
any new property must be private

stof commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -122,6 +123,7 @@ public function boot()
         }
 
         $this->booted = true;
+        $this->httpKernel = $this->getHttpKernel();
```
this looks wrong to be. It is perfectly possible to boot the kernel without needing the HttpKernel (perfect use case is the console)

gnugat replied:
Thanks for the feedback. Should I move it to `getHttpKernel()` then?

stof commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -133,8 +135,8 @@ public function terminate(Request $request, Response $response)
 
-        if ($this->getHttpKernel() instanceof TerminableInterface) {
-            $this->getHttpKernel()->terminate($request, $response);
+        if ($this->httpKernel instanceof TerminableInterface) {
```
the first change which could be done very easily would be to create a local variable with the HttpKernel here, to avoid calling it twice.

gnugat replied:
Agreed. Should I remove the property altogether then? This way we'd have one call in `handle` and one call here

gnugat committed:
Removed httpKernel class propoerty, created local httpKernel variable in terminate 

Taking @stof feedback into account

gnugat commented:
@sstok: good point for `__clone` and `shutdown`. Following @stof comments I've removed the class property entirely.
I'm sorry if I wasn't clear about my intentions: the usage of a profiler here was just highlighting the number of calls, not a performance issue. `http_kernel` being a shared service, we can call `container->get('http_kernel')` any number of time without worrying about performance (worst case scenario we have 2 function calls each time, with `getHttpKernel()`, which isn't much, especially since it's not done in a loop or similar).

It's just that since HttpKernel doesn't change I found it a bit silly to call `getHttpKernel()` many times, especially in `terminate`, when we could store it in a local variable (or a class property if we want to share it with `handle`).

gnugat committed:
Fixed typo

gnugat committed:
Fixed typo

Commits will be squashed before merge

gnugat committed:
Removing `$kernel->getHttpKernel()` spies

Because `$httpKernel->terminate()` spies are enough to know if we delegated termination

fabpot commented:
Tests do not pass as `$kernel` is not defined anymore in the tests. Closing as it does not bring anything useful anyway (and can complicate merges).

stof commented:
Well, the issue is that the changes done in the tests are wrong. ``getHttpKernel`` is still expected to be called

fabpot commented:
Of course, my point is that it is just not worth it. We've all already spent too much time on something that won't make any difference.
